File: Assign1/task3.scala
object task3_1{
        def main(args:Array[String]):Unit={
        val conf=new SparkConf().setAppName("spark-test").setMaster("local[*]")
        val sc=new SparkContext(conf)
        val textFile=sc.textFile(args(0).trim())

        val row=textFile.map(row=>parse(row))
        val n=args(4).trim().toInt
        val partion_no=args(3).trim().toInt
        var review_json=row.map(x=>((x \ "business_id").values.toString,1))
        if(args(2).trim()=="customized"){
        review_json=review_json.partitionBy(new HashPartitioner(partion_no)).persist()
        }

        val ans=Map(
        "n_partitions"->review_json.getNumPartitions,
        "n_items"->review_json.glom().map(x=>x.length).collect(),
        "result"->review_json.reduceByKey((x,y)=>x+y).filter(x=>x._2.toInt>n).map(x=>List(x._1,x._2)).collect())
        val file=new File(args(1).trim())
        val bw=new BufferedWriter(new FileWriter(file))

        bw.write(Json(DefaultFormats).write(ans).replace(":",": ").replace(",",", "))
        bw.close()
        sc.stop()

        }

        }

File: Assign1/task3.py
from pyspark import SparkContext
import json
import datetime
import sys

sc1 = SparkContext.getOrCreate()
review_json = sc1.textFile(sys.argv[1]).map(lambda k: json.loads(k)) 
n = int(sys.argv[5])
partion_no = int(sys.argv[4])
review_json  =review_json.map(lambda x:[x['business_id'],1] )
if(sys.argv[3].strip()== "customized") :
    review_json = review_json.partitionBy(int(partion_no), lambda key : hash(key)).persist()
part = review_json.getNumPartitions()
items =  review_json.glom().map(lambda x : len(x)).collect() 

ans = {
    "n_partitions" : part,
    "n_items" :  items,
    "result": review_json.reduceByKey(lambda x,y : x+y).filter(lambda x : x[1]> n).collect()
}
f = open(sys.argv[2],"w")
json.dump(ans, f)
f.close()
sc1.stop()